README.md
# Multi-channel contexts

Show and hide pages and content elements based on configurable "contexts".
With the use of contexts, TYPO3 is able to do multichannel output.

Examples for contexts:

- Screen size < 500px
- Browser runs on a tablet or mobile phone
- Location is 15km around a certain place
- User is from one of certain countries
- User entered website with GET-Parameter ``affID=foo``
- User IP matches a given rule
- A session variable is set
- A combination of any other rules

Apart from the context rules, this extension also provides an API to use
contexts in your own extensions.


<!-- TOC -->
* [Multi-channel contexts](#multi-channel-contexts)
  * [Setup](#setup)
  * [Creating and using contexts](#creating-and-using-contexts)
    * [Creating a context](#creating-a-context)
    * [Using a context](#using-a-context)
  * [Context types](#context-types)
    * [Domain](#domain)
      * [Domain matching](#domain-matching)
    * [GET parameter](#get-parameter)
    * [IP address](#ip-address)
    * [HTTP header](#http-header)
    * [Logical context combination](#logical-context-combination)
  * [Fluid template implementation](#fluid-template-implementation)
  * [TypoScript implementation](#typoscript-implementation)
<!-- TOC -->


## Setup
1. Install and activate extension ``contexts``
2. Clear TYPO3 cache

Optional: Install extensions ``contexts_geolocation`` for location-based
context rules (continent, country, area) and
``contexts_wurfl`` for device-based rules
(type check: phone, tablet, TV, ...; screen sizes, device/browser type).


## Creating and using contexts

### Creating a context
1. Log into the TYPO3 backend as administrator
2. Goto Web/List view, root page (ID 0)
3. Create a new record: TYPO3 contexts -> Context

   - Give it a title, e.g. "Affiliate ID: foo"
   - Select a type: "GET parameter"
   - GET parameter name: ``affID``
   - Parameter value: ``foo``
   - Activate "Store result in user session"
   - Save and close

### Using a context
1. Goto Web/Page, select a page
2. Edit a content element
3. Select the "Contexts" tab
4. For Context "Affiliate ID: foo", select "Visible: yes"
5. Save

View the page. The content element is not visible.

Now add ``?affID=foo`` to the URL and load it.
The content element will be visible now.

You can visit other pages now. When you come back, the content element
will still be visible - even though the GET parameter is not in the URL
anymore - because "Store result in user session" had been activated.


## Context types
The ``contexts`` extension ships with a number of simple contexts.
All of them get stored in table ``tx_contexts_contexts``.

### Domain
A domain context matches when the domain the user visits is in the
configured list.

This is helpful if the site is available on several domains, or
when it is deployed on development/stage/live systems - you may choose
to show a content element on the development system only.

#### Domain matching
You may use one domain per line.

When the domain does not begin with a dot, it will only match fully:
``www.example.org`` will not match the configured domain ``example.org``.

It is possible to use a dot in front of the domain name.
In this case, all subdomains will match:
``some.www.example.org`` matches the configured domain ``.example.org``.


### GET parameter
Checks if a GET parameter is available and has a certain value.

Activate "Store result in user session" to keep the context when navigating
between pages.

When leaving the parameter values field empty, any non-empty parameter value
will activate the context.


### IP address
Matches the user's IP address. IPv4 and IPv6 are supported.

Supported notations:

- Full addresses: ``80.76.201.32``
- Prefix: ``80.76.201.32/27``, ``FE80::/16``
- Wildcards: ``80.76.201.*``, ``80.76.*.37``, ``80.76.*.*``


### HTTP header
Checks if an HTTP header is available and has a certain value.

Activate "Store result in user session" to keep the context when navigating
between pages.

When leaving the parameter values field empty, any non-empty parameter value
will activate the context.


### Logical context combination
Combines other contexts with logical operators.

Contexts are referenced via their alias and can be combined with
the following signs:

- logical and: ``&&``
- logical or: ``||``
- negation: ``!``
- parentheses to group parts of expressions: ``(...)``


### Session variable
This context checks if a session variable with the given name is
set (is not NULL).


## Fluid template implementation
The implementation of a context query in fluid templates looks like::

```html
<div xmlns="http://www.w3.org/1999/xhtml" xmlns:contexts="http://typo3.org/ns/Tx_Contexts_ViewHelpers">
    <f:if condition="{contexts:matches(alias:'mobile')}">
        <f:then>is Mobile</f:then>
        <f:else>is not Mobile</f:else>
    </f:if>
</div>
```

## TypoScript implementation
The implementation of a context query in TypoScript looks like::

```typo3_typoscript
[contextMatch("mobile")]
    # do something, it's a mobile browser
[END]
```


ChangeLog
2016-04-21  Tobias.Hein  <tobias.hein@netresearch.de>

	* Version: 0.5.0

2016-04-18  Thomas Schöne  <thomas.schoene@netresearch.de>

	* TYPO-4217: pay attention on additionalCacheHashParams when calculating the cache hash

2016-03-15  Tobias Hein  <tobias.hein@netresearch.de>

	* MYC-869: use getIndpEnv to determine REMOTE_ADDR, allow reverse proxys to work

2016-01-06  Michael Schams  <typo3@2016.schams.net>

	* Implement new context type: HTTP header

	* Update documentation

2015-07-10  Christian Weiske  <christian.weiske@netresearch.de>

	* Fix #65964: Keep context configuration when copying content
      elements or page records

	* Version 0.4.3

2015-06-18  Tobias Hein  <tobias.hein@netresearch.de>

	* Check if tfse is a instance of tslib_fe

	* Version: 0.4.2

2014-07-31  Christian Weiske  <christian.weiske@netresearch.de>

	* Fix !negation in logical context combinations

	* Version 0.4.1

2014-03-21  Christian Weiske  <christian.weiske@netresearch.de.de>

	* Fix PHP warning when saving content element

2014-03-07  Christian Weiske  <christian.weiske@netresearch.de>

	* Prettify contexts list in TCA (page/content records)

	* Show message when no contexts defined

	* Version 0.4.0

2014-02-28  Christian Weiske  <christian.weiske@netresearch.de>

	* Fix bug: GET parameters get appended to all URLs even when
      they are stored in session

	* Set minimum TYPO3  version to  4.5.0

2014-02-28  Mathias Brodala  <mbrodala@pagemachine.de>

	* Implement #56269: Backport TSFE getHash hook to TYPO3 4.5

2013-11-22  Christian Weiske  <christian.weiske@netresearch.de>

	* Fix bug in contexts that caused DAM not to show any usage references

	* Fix a broken unit test

	* Remove manual.sxw; it is not needed by TER anymore

	* Set minimum TYPO3  version to  4.6.0

	* Version 0.3.1

2013-08-16  Christian Weiske  <christian.weiske@netresearch.de>

	* Support "extend to subpages" in 4.7, 6.0, 6.1 and 6.2

	* Fix fatal error in TYPO3 6.2

	* Version 0.3.0

2013-06-03  Christian Weiske  <christian.weiske@netresearch.de>

	* First public version with Domain, GetParam, IP, Combination support

	* Version 0.2.0

2013-01-30  Christian Opitz  <christian.opitz@netresearch.de>

	* Initial code generated with kickstarter


ext_emconf.php
<?php

/***************************************************************
 * Extension Manager/Repository config file for ext "contexts".
 *
 * Auto generated 21-04-2016 07:58
 *
 * Manual updates:
 * Only the data in the array - everything else is removed by next
 * writing. "version" and "dependencies" must not be touched!
 ***************************************************************/

$EM_CONF[$_EXTKEY] = array(
	'title' => 'Multi-channel contexts',
	'description' => 'Contexts for TYPO3 - can be used to modify page behaviour in the frontend based on several conditions',
	'category' => 'misc',
	'author' => 'Andre Hähnel, Christian Opitz, Christian Weiske, Marian Pollzien, Rico Sonntag',
	'author_email' => 'typo3.org@netresearch.de',
	'shy' => '',
	'dependencies' => '',
	'conflicts' => '',
	'priority' => 'top',
	'module' => '',
	'state' => 'alpha',
	'internal' => '',
	'uploadfolder' => 0,
	'createDirs' => '',
	'modify_tables' => '',
	'clearCacheOnLoad' => 1,
	'lockType' => '',
	'author_company' => 'Netresearch GmbH & Co.KG',
	'version' => '0.5.0',
	'constraints' => array(
		'depends' => array(
			'typo3' => '4.5.0-6.2.99',
		),
		'conflicts' => array(
		),
		'suggests' => array(
			'contexts_geolocation' => '',
			'contexts_wurfl' => '',
		),
	),
	'_md5_values_when_last_written' => 'a:59:{s:9:"build.xml";s:4:"64f8";s:9:"ChangeLog";s:4:"4c6e";s:21:"class.ux_tslib_fe.php";s:4:"5fa3";s:16:"ext_autoload.php";s:4:"d8e2";s:16:"ext_contexts.php";s:4:"3b91";s:12:"ext_icon.gif";s:4:"d6c7";s:17:"ext_localconf.php";s:4:"f1d4";s:14:"ext_tables.php";s:4:"8dc2";s:14:"ext_tables.sql";s:4:"087f";s:29:"icon_tx_contexts_contexts.gif";s:4:"d6c7";s:20:"overlay-contexts.png";s:4:"9252";s:10:"README.rst";s:4:"c4ef";s:12:"Settings.yml";s:4:"ae55";s:7:"tca.php";s:4:"6b8b";s:19:"Classes/Backend.php";s:4:"b19a";s:21:"Classes/Exception.php";s:4:"b002";s:29:"Classes/Api/Configuration.php";s:4:"6878";s:30:"Classes/Api/ContextMatcher.php";s:4:"4524";s:22:"Classes/Api/Record.php";s:4:"6327";s:28:"Classes/Context/Abstract.php";s:4:"39fa";s:29:"Classes/Context/Container.php";s:4:"a668";s:27:"Classes/Context/Factory.php";s:4:"14cb";s:27:"Classes/Context/Setting.php";s:4:"0e31";s:36:"Classes/Context/Type/Combination.php";s:4:"3468";s:32:"Classes/Context/Type/Default.php";s:4:"2541";s:31:"Classes/Context/Type/Domain.php";s:4:"4432";s:33:"Classes/Context/Type/GetParam.php";s:4:"0fdb";s:35:"Classes/Context/Type/HttpHeader.php";s:4:"f487";s:27:"Classes/Context/Type/Ip.php";s:4:"149c";s:32:"Classes/Context/Type/Session.php";s:4:"bb72";s:63:"Classes/Context/Type/Combination/LogicalExpressionEvaluator.php";s:4:"9234";s:73:"Classes/Context/Type/Combination/LogicalExpressionEvaluator/Exception.php";s:4:"582b";s:45:"Classes/Context/Type/GetParam/TsfeService.php";s:4:"3107";s:24:"Classes/Service/Icon.php";s:4:"831d";s:27:"Classes/Service/Install.php";s:4:"957e";s:24:"Classes/Service/Page.php";s:4:"1643";s:23:"Classes/Service/Tca.php";s:4:"3c86";s:27:"Classes/Service/Tcemain.php";s:4:"bc8d";s:24:"Classes/Service/Tsfe.php";s:4:"40e2";s:41:"Classes/ViewHelpers/MatchesViewHelper.php";s:4:"b6dd";s:50:"Configuration/flexform/ContextType/Combination.xml";s:4:"a821";s:45:"Configuration/flexform/ContextType/Domain.xml";s:4:"9b14";s:47:"Configuration/flexform/ContextType/GetParam.xml";s:4:"40b6";s:49:"Configuration/flexform/ContextType/HttpHeader.xml";s:4:"4699";s:41:"Configuration/flexform/ContextType/Ip.xml";s:4:"0d38";s:46:"Configuration/flexform/ContextType/Session.xml";s:4:"b5db";s:24:"Library/ts_connector.php";s:4:"952c";s:39:"Resources/Private/Language/flexform.xml";s:4:"388c";s:43:"Resources/Private/Language/locallang_db.xml";s:4:"ff3e";s:37:"Resources/Private/csh/Combination.xml";s:4:"0601";s:34:"Resources/Public/StyleSheet/be.css";s:4:"1db5";s:19:"tests/bootstrap.php";s:4:"4b70";s:17:"tests/phpunit.xml";s:4:"3704";s:18:"tests/TestBase.php";s:4:"7770";s:38:"tests/Classes/Context/AbstractTest.php";s:4:"cea2";s:46:"tests/Classes/Context/Type/CombinationTest.php";s:4:"0655";s:43:"tests/Classes/Context/Type/GetParamTest.php";s:4:"6e3c";s:37:"tests/Classes/Context/Type/IpTest.php";s:4:"0e00";s:73:"tests/Classes/Context/Type/Combination/LogicalExpressionEvaluatorTest.php";s:4:"73bd";}',
	'comment' => 'First public version',
	'suggests' => array(
	),
);

?>